### PR TITLE
Add blog post from Comcast

### PR DESCRIPTION
**Pull request by posulliv**

### CODE REVIEW

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+column, etc.,  ran these queries multiple times and execution times were
+consistent. This ruled out a potential problem with HDFS.
+
+Next, we took a closer look  at the portion of the slow running queries
+involving the A table and came up with the simplest possible query that could
+demonstrate the problem. We discovered that the following query performed as
+expected:
+
+```
+select
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:39]: For here and all statements .. use uppercase SQL keywords like in the docs

SELECT count(a.c) 
FROM
...

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+ count(a.c1)
+from
+ hive.schema1.A a, hive.schema2.B da
+where
+ a.day_id = da.date_id
+ and a.day_id BETWEEN '2021-03-22' AND '2021-04-21'
+```
+
+But adding a predicate caused the performance problem to appear:
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:48]: add in this sentence what the predicate is..

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+We narrowed the problem down to the `Scan/Filter/Project` operator using the
+output of `EXPLAIN ANALYZE` from Trino. For the query that performed as
+expected, this stage had the following CPU stats:
+
+```
+CPU: 2.39h, Scheduled: 4.47h, Input: 17434967615 rows (357.47GB)
+```
+
+For the simple version of the query performing poorly, the same stage has the
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:69]: well.. reword .. this is for the more complex version with the predicate right?

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+following CPU stats:
+
+```
+CPU: 3.73d, Scheduled: 48.01d, Input: 17052985227 rows (413.98GB)
+```
+
+This shows that for roughly the equivalent amount of data, Trino used
+significantly more CPU (3.73 days to 2.39 hours!!). Our next step was to
+determine why that was and exactly what was using additional CPU.
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:78]: was to determine possible reasons.

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+We generated a few jstack and JFR profiles of the Starburst Java process from
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:80]: link to jstack maybe 

and what is JFR (spell it out)

And 

of the Trino Java process

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+one of the worker nodes while the scan stage was running. After analyzing these
+profiles, we found no obvious problem; Trino performed as expected.
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:82]: problem. Trino

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+Next, we looked at the list of tasks in the web UI to see what the distribution
+of CPU times for each stage was:
+
+![](/assets/blog/jvm-issues-at-comcast/web_ui_before.png)
+
+Notice in the above screenshot, some workers have tasks that only use up a few
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:89]: You can leave out the "Notice in the above screenshot"

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+minutes of CPU time and others have tasks that use up to 2 hours of CPU time!
+Different query runs would show this would happen to different workers so it
+was not a problem with any one individual worker.
+
+We discussed this with Starburst engineer, [Piotr Findeisen](https://github.com/findepi),
+and came to the conclusion that this could potentially be an issue with JVM
+code deoptimization. After re-compiling a method a certain number of times,
+the JVM will refuse to do so any more and will run the method in interpreted
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:97]: change to present tense

the JVM refuses .. and runs in the much slower interpreted mode.

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+mode, which is much slower.
+
+The evidence for this is what we highlighted above: that the CPU used by the
+same tasks on different workers vary by a factor of approximately 30. This is
+the typical difference one would expect for compiled versus interpreted code,
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:102]: remove "one would expect"

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+* `-XX:PerBytecodeRecompilationCutoff=10000`
+
+These settings increased the recompilation cutoff limit. They are now also
+included in the default `jvm.config` settings that ship with Trino since the
+348 release. 
+
+Since we have been running Trino in production for some time, we did not have
+these settings in our `jvm.config`.
+
+### Initial Results
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:118]: Initial results

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+Execution time observed  with the JVM options in place was 4 minutes and 51
+seconds. The CPU stats for the scan/filter/project stage for this query now
+look like:
+
+```
+CPU: 3.22h, Scheduled: 7.21h, Input: 17631445897 rows (428.03GB)
+```
+
+The CPU used by individual tasks is much more uniform as shown in the following
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:128]: is much more uniform:

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+screenshot from the web UI:
+
+![](/assets/blog/jvm-issues-at-comcast/web_ui_after.png)
+
+## Code Cache
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:133]: Code cache

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+We noticed that the cluster’s overall CPU utilization decreased after the
+cluster was up for a few days and there would be a few workers where tasks
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:136]: few days, and there

File: _posts/2021-10-06-jvm-issues-at-comcast.md [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+SELECT
+    node,
+    regexp_extract(usage, 'max=(-?\d*)', 1) as max,
+    regexp_extract(usage, 'used=(-?\d*)', 1) AS used
+FROM
+  jmx.current."java.lang:name=codeheap 'non-profiled nmethods',type=memorypool"
+ORDER BY used DESC
+```
+
+## Off Heap Memory Usage
```
Comment (mosabua) [inline _posts/2021-10-06-jvm-issues-at-comcast.md:202]: Off heap memory usage

**Review by mosabua — CHANGES_REQUESTED** [review]
A bunch of nits and suggestions. 

Beyond that I think you need some sort of Summary at the end.

And also .. can we also get a user testimonial and logo for the site? I can help with details or even send the PR if you supply text, URL and logo. See https://trino.io/users.html
cc @bitsondatadev @martint

**Comment by bitsondatadev** [comment]
@mosabua just so it's clear @posulliv is a support engineer who worked with Comcast but it would be cool if Comcast could add their testimonial. 

Padraig can you speak with David and Saj to see if they would be interested in adding the Comcast logo and testimonial on the Trino site?

**Comment by mosabua** [comment]
I am of course aware of the role of @posulliv ...

**Comment by posulliv** [comment]
@mosabua thank you for the review and many comments! I went through and applied them all. I also added a brief summary at the end.

**Comment by posulliv** [comment]
I'll check with the Comcast folks about a user testimonial on the Trino site and let you know.

**Comment by posulliv** [comment]
@mosabua @bitsondatadev is there anything else you need on my end for this post?

**Comment by mosabua** [comment]
Its up to @martint @dain @electrum and other maintainers to decide on next steps. I can review again if desired.

**Pull request merged**